modules/core/src/main/java/org/apache/ignite/internal/causality/AbstractVersionedValue.java

```diff
@@ -0,0 +1,359 @@
+                } else {
+                    if (previousCompleteFuture == null) {
+                        future.complete(getDefault());
+                    } else {
+                        copyState(previousCompleteFuture, future);
+                    }
+
+                    return future;
+                }
+            });
+
+            // Delete older tokens if their amount exceeds the configured history size.
+            trimHistory();
+        } finally {
+            readWriteLock.writeLock().unlock();
+        }
+
+        notifyCompletionListeners(causalityToken, futureForToken);
+    }
+
+    /**
+     * Completes the Versioned Value for the given token with the given {@code future}. This method will look for the previous complete
+     * token and complete all registered futures in the {@code (prevToken, causalityToken)} range. If no {@code complete} methods have
+     * been called before, all these futures will be complete with the configured default value.
+     *
+     * <p>Calling this method will trigger the {@link #whenComplete} listeners for the given token.
+     */
+    protected void complete(long causalityToken, CompletableFuture<T> future) {
+        assert future.isDone();
+
+        readWriteLock.writeLock().lock();
+
+        try {
+            setActualToken(causalityToken);
+
+            completePreviousFutures();
+
+            CompletableFuture<T> existingFuture = history.putIfAbsent(causalityToken, future);
+
+            if (existingFuture != null) {
+                copyState(future, existingFuture);
+            }
+
+            // Delete older tokens if their amount exceeds the configured history size.
+            trimHistory();
+        } finally {
+            readWriteLock.writeLock().unlock();
+        }
+
+        notifyCompletionListeners(causalityToken, future);
+    }
+
+    /**
+     * Completes a future related with a specific causality token. It is called only on storage revision update.
+     */
+    @Nullable
+    private CompletableFuture<T> completePreviousFutures() {
+        NavigableMap<Long, CompletableFuture<T>> headMap = history.headMap(actualToken);
+
+        if (headMap.isEmpty()) {
+            return null;
+        }
+
+        List<CompletableFuture<T>> futuresToComplete = List.of();
+
+        CompletableFuture<T> previousCompleteFuture = null;
+
+        for (CompletableFuture<T> future : headMap.descendingMap().values()) {
+            if (future.isDone()) {
+                previousCompleteFuture = future;
+
+                break;
+            }
+
+            // Lazy initialization.
+            if (futuresToComplete.isEmpty()) {
+                futuresToComplete = new ArrayList<>();
+            }
+
+            futuresToComplete.add(future);
+        }
+
+        // We found the first complete future, but there are no incomplete futures up to the actual token, so there's nothing to do.
+        if (futuresToComplete.isEmpty()) {
```

Comment:
> Seems counter-intuitive, that the value is not updated, but the future is not completed at the same time

Can you elaborate please?

Reply:
Ok, imagine that you have VV completed on version 1, and some code waits for versions 2 and 3.
But you only complete VV on version 4. 2 and 3 will have the same value as 1, but, if we knew that in advance, why did we have to wait for 4 to complete them?
This part is counter-intuitive, I would probably expect the assertion that the collection is empty, and that our restrictions are very strict.

Reply:
> but, if we knew that in advance, why did we have to wait for 4 to complete them?

How do we know this in advance when you haven't observed the 2 and 3 revision explicitly (that's what the `complete(long)` method for)?

But I don't think this can happen in the production code, however, because most VersionedValues are `IncrementalVersionedValues` which now get notified of every revision....

Reply:
Ok, if that's intentional and people expected that, what can I do